File: python_side/complex_snippets.py
# TODO Normalize and add to rsnips
# subsetting on a particular column with a filter
df['Age'][df['Survived'] == 1]
df['Isalone'].loc[df['FamilyMembers'] > 1]
# subsetting on main dataframe then referencing a column(s)
df[vd['PassengerId'] == 1284]['Survived']
df[DataSetTrain['Survived'] == 0][['Name', 'Sex', 'Age', 'Pclass', 'Cabin']]
# subsetting using which with chained conditions where rhs is either numeric or string
df[(df.Pclass == 3) & (df.Embarked == 'S') & (df.SibSp == 0) & (df.Parch == 0) & (df.Sex == 'male') & (df.Age > 50.0)]
# subsetting using chained conditions using mix of boolean operators?

# subsetting with a mix of a functions inside [] (notnull, isnull, isin)
df['Cabin'][df['Cabin'].notnull()].head()
df.loc[(df.Fare.isnull()), :]
x_test.loc[x_test['Fare'].isnull()]['Pclass']
df[fulldata.Fare.isnull()][['Fare', 'Pclass', 'Embarked']]
df[df.Fare.isnull() == True]
df[(df['NoPerTicket'] == 1) & (df['Age'].isnull() == False) & ((df['Parch'] > 0) | (df['SibSp'] > 0))]
df[df.Title.isin(['Rev', 'Dr', 'Col', 'Capt', 'Major']) & (df.Sex != 'male')]
df[df.Name.isin(['Connolly, Miss. Kate', 'Kelly, Mr. James'])].sort_values(by='Name')

# function call on subsetting operation(s)
df[numeric_variables].head(3)
df[(df['Embarked'] == 'C') & (df.Fare > 200)].head()
df[(df.Sex == 'female') & (df.Pclass == 3) & (df.Survived == 1)].head(20)
df[['Age', 'Age_0', 'Age_1', 'Age_2', 'Age_3', 'Age_4', 'Age_5', 'Age_6']].head(10)
df[(df['Age'] > 8.0) & (df['Age'] <= 10.0)].sort_values('Age')
df.loc[df['SharedTicket'] == 1, 'Ticket'].sort_values()
df[df.Age.notnull()].query("Title =='MASTER' and Age >= 10")

# chained function calls
df.sort_values(by='Name').head(10)
df.sort_values(by='SibSp').describe(percentiles=[0.68, 0.69])
df.drop(categorical_columns, axis=1).describe(include=[np.number])
df.query('CabinCount > 1').head()

# computation using dimensions of two dataframes
df[(df.Fare == 0) & (df.Pclass == 2)].shape[0] / (1.0 * df[df.Fare == 0].shape[0])

# inverse boolean vector using ~ (contrast to R, ~ equivalent to ! is higher in precedence so parens are needed)
# for e.g
"""
>>> df
   a  b
0  1  4
1  2  5
2  3  6
>>> ~df.a > 1
0    False
1    False
2    False
Name: a, dtype: bool
>>> ~(df.a > 1)
0     True
1    False
2    False
Name: a, dtype: bool
"""
df[~((df.Embarked == 'S') | (df.Embarked == 'C') | (df.Embarked == 'Q'))]


# Stats
# summary on a chained subsetting operation
df[df['Pclass'] == 3]['Fare'].describe()
train.loc[train['Pclass'] == 3]['Deck'].describe()

File: r_side/complex_snippets.R
# TODO Normalize and add to rsnips
# subsetting on a particular column with a filter
full_data$Title[full_data$Title == 'Mlle'] 
# subsetting on main dataframe then referencing a column
full[is.na(full$embarked), ]$embarked <- 'S'
full[is.na(full$Age) & full$Title == "Mr", ]$Age
# subsetting using which with chained conditions where rhs is either numeric or string
train[which(train$Parch > 0 & train$Age > 18 & train$Name == "Mrs."), c("Mother")] <- 1
full$Mother[full$Sex == 'female' & full$Parch > 0 & full$Age > 18 & full$Title != 'Miss'] <- 'Mother'
# subsetting using chained conditions using mix of boolean operators
fulldata$isKids[fulldata$Age <=18  & (fulldata$Parch > 0 | fulldata$TicketGroupSize>1 | fulldata$FamilynameSize>1)  & fulldata$Title != 'Mrs'] <- 1


# subsetting with a mix of a function and regular condition (which, is.na, grepl, %in%)
train$farediv[which(train[, 10] > 20)] <- 2 
test$Age[is.na(test$Age) & test$Name == "Master."] <- masterage
train$Age[grepl("Master\\.", train$Name) & is.na(train$Age)] = mean.master
combi$Title[combi$Title %in% c('Mlle','Mme')]<-'Mlle'
combined.df$TitleGroup[which(combined.df$Title %in% c("Rev"))] <- "MrRev" 

# subsetting while retaining the original dimension (keeping it a data frame)
data[[i]] <- full[train_dev_idx, var[[i]], drop=F]

# function call on subsetting operation(s)
head(whole.data)
head(full_data$Name)
head(train_data[,coloumns.to.fit],10)
head(data[c("FamilyName", "Sector")])
head(subset(df,fpr<0.2))


# chained function calls
age[age_na_rows] = sample(na.omit(full$Age), length(age_na_rows))
head(subset(cutoffs, fpr < 0.2))


# inverse boolean vector using ! (note: the ! has lower precedence than the other boolean operators)
# this is in contrast to Python where the equivalent ~ is higher precedence. 
# example in R:
# > df
#   a b
# 1 1 4
# 2 2 5
# 3 3 6
# > !df$a > 1
# [1]  TRUE FALSE FALSE
# > !(df$a > 1)
# [1]  TRUE FALSE FALSE
train$Cabin[which(!train$Cabin == "")]<-1

# 
trn_m$Survived_F = factor(trn_m$Survived)
# renaming for a column where there are no missing values
full[which(is.na(full$Survived)), 2] <- 0

# Stats stuff
# double $
summary(train_imp$imp$Age)
summary(df[, c('Pclass', 'Survived')])

# na.rm is F by default whereas pandas is T by default
mean(my.test.data[,i], na.rm = TRUE)
full$Fare[1044] <- median(full[full$Pclass == 3 & full$Embarked == 'S', ]$Fare,na.rm = TRUE)

# More complex stuff that transforms columns
# creation/transformation of column based on 1 other column
mutate(rate=n/FamilySize)
# creation/transformation of column based on 2 other columns
mutate(FamSize = SibSp + Parch)